crates/noirc_evaluator/src/ssa/opt/mem2reg.rs

```diff
@@ -1,378 +1,501 @@
         let post_order = PostOrder::with_function(function);
         let dom_tree = DominatorTree::with_cfg_and_post_order(&cfg, &post_order);
+
         PerFunctionContext {
-            last_stores_with_block: BTreeMap::new(),
-            load_values_to_substitute_per_func: BTreeMap::new(),
-            store_ids: Vec::new(),
             cfg,
             post_order,
             dom_tree,
-            loops: find_all_loops(function),
+            blocks: BTreeMap::new(),
+            instructions_to_remove: BTreeSet::new(),
         }
     }
-}
 
-/// An AllocId is the ValueId returned from an allocate instruction. E.g. v0 in v0 = allocate.
-/// This type alias is used to help signal where the only valid ValueIds are those that are from
-/// an allocate instruction.
-type AllocId = ValueId;
+    /// Apply the mem2reg pass to the given function.
+    ///
+    /// This function is expected to be the same one that the internal cfg, post_order, and
+    /// dom_tree were created from.
+    fn mem2reg(&mut self, function: &mut Function) {
+        // Iterate each block in reverse post order = forward order
+        let mut block_order = PostOrder::with_function(function).into_vec();
+        block_order.reverse();
+
+        for block in block_order {
+            let references = self.find_starting_references(block);
+            self.analyze_block(function, block, references);
+        }
+    }
 
-impl PerFunctionContext {
-    // Attempts to remove load instructions for which the result is already known from previous
-    // store instructions to the same address in the same block.
-    fn analyze_allocations_and_eliminate_known_loads(
-        &mut self,
-        dfg: &mut DataFlowGraph,
-        loads_to_substitute: &mut BTreeMap<InstructionId, ValueId>,
-        load_values_to_substitute_per_block: &mut BTreeMap<ValueId, ValueId>,
-        block_id: BasicBlockId,
-    ) -> HashSet<AllocId> {
-        let mut protected_allocations = HashSet::new();
-        let block = &dfg[block_id];
-
-        // Check whether the block has a common successor here to avoid analyzing
-        // the CFG for every block instruction.
-        let has_common_successor = self.has_common_successor(block_id);
-
-        for instruction_id in block.instructions() {
-            match &dfg[*instruction_id] {
-                Instruction::Store { mut address, value } => {
-                    if has_common_successor {
-                        protected_allocations.insert(address);
-                    }
-                    address = self.fetch_load_value_to_substitute(block_id, address);
+    /// The value of each reference at the start of the given block is the unification
+    /// of the value of the same reference at the end of its predecessor blocks.
+    fn find_starting_references(&mut self, block: BasicBlockId) -> Block {
+        let mut predecessors = self.cfg.predecessors(block);
+
+        if let Some(first_predecessor) = predecessors.next() {
```

Comment:
Is there is an issue with unifying values across multiple predecessors? Should we add a check that `first_predecessor` actually dominates `block` ?

Reply:
I'm not sure what this comment means - we do unify values across multiple (direct) predecessors. The domination check is unneeded since all predecessors do not need to dominate the block.

Reply:
Ah I misread how this method was going over the predecessors, I see now. And makes sense with the domination check you are correct.